<head>
    <meta name="keywords" content="CNV, integrative analysis, NGS, array CGH,
          multidimensional data">
    <title> genomeCAT </title>
    <style>
        p {font-family: Verdana,Arial,Helvetica,sans-serif;}
        body {color: darkslategray; background-color: teal; 
              font-family: Verdana,Arial,Helvetica,sans-serif; font-size: small}
        table { background-color: white; border-spacing: 20px;  margin: 1px; width: 800px}
        td,tr { vertical-align: text-top ; padding: 10px;}
        a:link {color: darkred; font-weight: bold}
        td.text { background-color: whitesmoke; width: 150px; border-bottom:  red dotted 1px}
        div.subline {padding-bottom:  10px; padding-top: 10px; border-bottom:  dotted 1px}
        header { background-color:   darkturquoise; padding: 10px; font-size:  20px; font-weight: bold;   }

        aside { margin: 1px; float:left; width: 150px;  }
        section {float: right; margin: 1px;  }

		
        nav {background-color: powderblue; font-weight: bold;
             font-size:  medium; float: right; margin: 1px; width: 20% }


        div.left {padding: 10px; background-color: whitesmoke; position: absolute; bottom: 3%;  
                  vertical-align:text-top; width: 150px}
		
		#topnava{			
			letter-spacing: 3px;
			font-size:  10px;
			vertical-align:bottom;
			}

		#topnava ul {
			list-style-type:none;
			margin:0;
			padding:0;
			}			

		#topnava li {
			padding: 0 6px 0 8px;
			border-left: 1px solid #c6cbdd;
			display:inline;
			}

		#topnava a {
			margin: 0;
			padding: 2px;
			text-transform: capitalize;
			text-decoration:none;
			}

		#topnava a:hover {
			background-color: powderblue;;
			text-decoration:none;
			}

    </style>
</head>
<body>  
    <header>
        <div>GenomeCAT - a versatile tool for the analysis and integrative visualization of DNA copy number variants<br/><br/></div> 
	 
		<div id="topnava">
		<ul>
		<li><a href="http://genomecat.github.io/genomeCATSuite" title="Home">Intro</a></li>
		<li><a href="http://genomecat.github.io/genomeCATSuite/download.html" title="Download">Download</a></li>
		<li><a href="http://genomecat.github.io/genomeCATSuite/doc.html" target="_self" title="DOCUMENTATION">Documentation</a></li>
		<li><a href="https://github.com/genomeCAT/genomecatPub/raw/master/resources/faqs.html" target="self" title="FAQs" target="_self">FAQs</a></li>
		<li><a href="https://github.com/genomeCAT/genomeCATSuite" title="Source Repository" target="_blank">Source Repository</a></li>
		</ul></div>
	</header>
    
  
    <aside>
        
        <div class="left"> 
		
		 
		  
            <a href="https://github.com/genomeCAT/genomeCATSuite" target="_blank">Source Repository</a> on GitHub.         
               
            
            <hr style="height: 10px;  border: none ">
            <a href="http://www.dfg.de/index.jsp"  target="_blank"  title="" >
                <img src="https://raw.githubusercontent.com/genomeCAT/genomecatPub/c29b0f218e40afbb1f2b572fae6d6b17dc01f01f/images/DFG_klein.jpg"
                                         alt="dfg"> 
            </a>
            <br>This project is funded by the DFG <br>(UL 342/2-1)

        </div>
    </aside>

    <nav>
        <div style="padding: 10px">
            <p style=" text-transform: uppercase">New features</p>         


            <div class="subline">
                + implementation of new R-packages for "NGS import"</div>

            <div class="subline">+ automatic search for software updates</div>

            <div class="subline">+ optimzed menu navigation</div>
            <div class="subline">+ export relative frequencies. <br/>
            </div>

            <div class="subline">+ import of data in BED, BedGraph, WIG format. <br/>
                Click <a href="https://github.com/genomeCAT/genomecatPub/blob/master/resources/BED_Import.zip?raw=true">
                    here</a> for a demonstration</div>
			
			<div class="subline"><br/><br/>
                            If you encounter any problems or have some comments that could help us to improve the software, 
                            please don't hesitate to contact <a href="mailto:tebel@molgen.mpg.de">Katrin Tebel</a> or 
                            <a href="mailto:ullmann@molgen.mpg.de">Reinhard Ullmann.</a></div>
			</div>
    </nav>



    <section> 
        <table>
            <colgroup span="3"></colgroup>

            <tr>
                <td colspan="3" >

                    <span>
                        <strong>Starting from <a href="http://www.biomedcentral.com/1471-2105/6/85">CGHPRO</a>, 
                            a software for the analysis and visualization of array CGH data, we are creating 
                            a program for the integrative analysis of genetic and epigenetic changes identified
                            by array and sequencing based experiment types, respectively.<br/>The full installation package can be downloaded 
							<a href="https://github.com/genomeCAT/genomecatPub/raw/master/updates/download.html">here</a>.
                            A video demonstrating the installation process can be downloaded 
                            <a href="https://github.com/genomeCAT/genomecatPub/blob/master/resources/HowToInstalGenomeCAT.zip?raw=true">here</a>.
                            </strong></span>                          

                </td>
            </tr>


            <tr style="text-align: center">
                <td>
                    <a  href="https://raw.githubusercontent.com/genomeCAT/genomecatPub/master/images/single_view.jpg" title="">
                        <img src="https://raw.githubusercontent.com/genomeCAT/genomecatPub/master/images/teaser_single_view.jpg" height="188" alt="single_view" width="143" />
                    </a>
                </td>
                <td>
                    <a  href="https://raw.githubusercontent.com/genomeCAT/genomecatPub/master/images/cat_view.jpg" title="">
                        <img src="https://raw.githubusercontent.com/genomeCAT/genomecatPub/master/images/teaser_cat_view.jpg" height="188" alt="cat_view" width="143" />
                    </a>
                </td>
                <td>
                    <a  href="https://raw.githubusercontent.com/genomeCAT/genomecatPub/master/images/group_view.jpg" title="">
                        <img src="https://raw.githubusercontent.com/genomeCAT/genomecatPub/master/images/teaser_group_view.jpg" height="188" alt="single_view" width="143" />
                    </a>
                </td>
            </tr>
            <tr >
                <td class="text">

                    <span>
                        <strong>Single View</strong><br/>A tool to visualize and explore single 
                        array CGH experiments. 
                        A short video introducing the features of this tool can be downloaded 
                         <a href="https://github.com/genomeCAT/genomecatPub/raw/master/resources/single_view.zip?raw=true">here.</a>. 
						 Please also consult our <a href="https://github.com/genomeCAT/genomecatPub/raw/master/resources/SingleView.pdf" target="_blank">Manual</a>.
                    </span>


                </td>
                <td class="text">
                    <span>
                        <strong>Comparative View</strong><br/>
                        A tool to compare data from the same or different experiment types. 
                        Data can be averaged over genes, fixed genomic intervals or regions of 
                        interest to enable follow up statistical analysis (see 
						<a href="https://github.com/genomeCAT/genomecatPub/raw/master/resources/ComparativeView.pdf" target="_blank">Manual</a>).
                        A video demonstrating the workflow is available 
                        <a href="https://github.com/genomeCAT/genomecatPub/blob/master/resources/comparative_view.zip?raw=true">here</a>.
						</span>

                </td>
                <td class="text">

                    <span>
                        <strong>Group Explorer</strong><br />A tool to identify recurrent aberrations in
                        large cohorts and to compare patterns of chromosomal aberrations between groups of patients. 
                        A Video demonstrating the workflow can be downloaded 
						<a href="https://github.com/genomeCAT/genomecatPub/blob/master/resources/group_explorer.zip?raw=true">here</a>.</span>
						Please also consult our <a href="https://github.com/genomeCAT/genomecatPub/raw/master/resources/GroupExplorer.pdf" target="_blank">Manual</a>.
                   
                </td>
            </tr><!-- <tr>
                <td colspan="3" >

                    <span>
                        </span>                          

                </td>
            </tr>-->

        </table>
    </section>


</body>
